olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/provenance/io/ProvenanceMarshaller.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates.
+ *
+ * Licensed under the 2-clause BSD license.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.oracle.labs.mlrg.olcut.provenance.io;
+
+import com.oracle.labs.mlrg.olcut.provenance.ObjectProvenance;
+import com.oracle.labs.mlrg.olcut.provenance.ProvenanceUtil;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+/**
+ * Interface for serializing and deserializing marshalled provenances to
+ * and from strings or files.
+ */
+public interface ProvenanceMarshaller {
+
+    /**
+     * The file extension this ProvenanceMarshaller supports.
+     * @return The file extension.
+     */
+    public String getFileExtension();
```

Comment:
It looks like this isn't used anywhere that I could see.  Was the thought to be able to have a ProvenanceUtil (or something) method that could deserialize from a path by picking the right ProvenanceMarshaller based on extension? Or maybe this is just a placeholder for future shenanigans of that sort?

Reply:
Yeah it's to mirror the `ConfigurationManager`'s `FileFormatFactory` stuff in case we want to dynamically load them later. We probably should move this over to being an SPI but I've never written one of those and don't have time to figure it out before the release.

Reply:
Of course the SPI mechanism gets confused easily when making shadow jars (due to the use META-INF), requiring special casing, so avoiding it may not be the worst idea.